velox/functions/prestosql/aggregates/AverageAggregate.cpp

```diff
@@ -223,50 +236,72 @@ class AverageAggregate : public exec::Aggregate {
         baseRowVector->childAt(0)->template as<SimpleVector<TAccumulator>>();
     auto baseCountVector =
         baseRowVector->childAt(1)->template as<SimpleVector<int64_t>>();
-
-    if (decodedPartial_.isConstantMapping()) {
-      if (!decodedPartial_.isNullAt(0)) {
-        auto decodedIndex = decodedPartial_.index(0);
-        if constexpr (checkNullFields) {
-          VELOX_USER_CHECK(
-              !baseSumVector->isNullAt(decodedIndex) &&
-              !baseCountVector->isNullAt(decodedIndex));
-        }
-        auto count = baseCountVector->valueAt(decodedIndex);
-        auto sum = baseSumVector->valueAt(decodedIndex);
+    if (decodedPartial_.isIdentityMapping() && baseSumVector &&
+        baseCountVector) {
+      if (decodedPartial_.mayHaveNulls()) {
+        rows.applyToSelected([&](vector_size_t i) {
+          if (baseRowVector->isNullAt(i)) {
+            return;
+          }
+          if constexpr (checkNullFields) {
+            VELOX_USER_CHECK(
+                !baseSumVector->isNullAt(i) && !baseCountVector->isNullAt(i));
+          }
+          updateNonNullValue(
+              groups[i],
+              baseCountVector->valueAt(i),
+              baseSumVector->valueAt(i));
+        });
+      } else {
         rows.applyToSelected([&](vector_size_t i) {
-          updateNonNullValue(groups[i], count, sum);
+          if constexpr (checkNullFields) {
+            VELOX_USER_CHECK(
+                !baseSumVector->isNullAt(i) && !baseCountVector->isNullAt(i));
+          }
+          updateNonNullValue(
+              groups[i],
+              baseCountVector->valueAt(i),
+              baseSumVector->valueAt(i));
         });
       }
-    } else if (decodedPartial_.mayHaveNulls()) {
-      rows.applyToSelected([&](vector_size_t i) {
-        if (decodedPartial_.isNullAt(i)) {
-          return;
-        }
-        auto decodedIndex = decodedPartial_.index(i);
-        if constexpr (checkNullFields) {
-          VELOX_USER_CHECK(
-              !baseSumVector->isNullAt(decodedIndex) &&
-              !baseCountVector->isNullAt(decodedIndex));
-        }
-        updateNonNullValue(
-            groups[i],
-            baseCountVector->valueAt(decodedIndex),
-            baseSumVector->valueAt(decodedIndex));
-      });
     } else {
-      rows.applyToSelected([&](vector_size_t i) {
-        auto decodedIndex = decodedPartial_.index(i);
-        if constexpr (checkNullFields) {
-          VELOX_USER_CHECK(
-              !baseSumVector->isNullAt(decodedIndex) &&
-              !baseCountVector->isNullAt(decodedIndex));
+      SelectivityVector baseRows = translateToInnerRows(decodedPartial_, rows);
+
+      DecodedVector baseSumDecoded(*baseRowVector->childAt(0), baseRows);
+      DecodedVector baseCountDecoded(*baseRowVector->childAt(1), baseRows);
+
+      if (decodedPartial_.isConstantMapping()) {
```

Comment:
Wonder if it makes sense to have (decodedPartial_.isConstantMapping()) as the top level check. This will work for both flat and encoded children as we dont need to put the sum and count vector in a decoded vector to read the values. only 1 value needs to be read, therefore traversing multiple layers once for each should be ok and faster because we dont traverse all rows when creating a decoded vector.